core/iwasm/libraries/lib-wasi-threads/stress_test/mutex_common.h

```diff
@@ -0,0 +1,261 @@
+        if (call_stat.calls[i] != 0) {
+            assert(call_stat.calls[i] < NUM_ITER / NUM_THREADS
+                   && "Call distribution by TID is out of range");
+            assert(call_stat.calls[i] > NUM_ITER / (NUM_THREADS + 2)
+                   && "Call distribution by TID is out of range");
+        }
+    }
+}
+
+void
+print_stat()
+{
+    fprintf(stderr, "Thread calls count by TID\n");
+    for (int i = 0; i < NUM_THREADS + 1; ++i) {
+        if (call_stat.tids[i] != 0) {
+            fprintf(stderr, "TID: %d; Calls: %d\n", call_stat.tids[i],
+                    call_stat.calls[i]);
+        }
+    }
+}
+
+// xor with a random value usually gives quite a uniform distribution
+uint32_t
+hash(uint32_t counter)
+{
+    return (counter << 22) ^ (0xA2BC22A); // Random value for hash func
+}
+
+// Locking and unlocking the same thread
+void
+same_thread_test(pthread_mutex_t *mutex)
+{
+    assert(pthread_mutex_lock(mutex) == 0
+           && "Main thread should be able to lock a mutex");
+    assert(pthread_mutex_unlock(mutex) == 0
+           && "Main thread should be able to unlock a mutex");
+}
+
+void *
+inc_until_limit(void *arg)
+{
+    MutexCounter mutex_counter = *(MutexCounter *)(arg);
+    int sleep_us = 0;
+    while (!pthread_mutex_lock(mutex_counter.mutex)
+           && *mutex_counter.counter < mutex_counter.limit) {
+        (*mutex_counter.counter)++;
+        add_to_stat((int)(pthread_self()));
+        if (mutex_counter.is_sleeping) {
+            sleep_us = hash(*mutex_counter.counter) % 1000;
+        }
+
+        assert(pthread_mutex_unlock(mutex_counter.mutex) == 0
+               && "Should be able to unlock a mutex");
+        if (mutex_counter.is_sleeping) {
+            usleep(sleep_us);
+        }
+    }
+
+    assert(*mutex_counter.counter == mutex_counter.limit);
+    assert(pthread_mutex_unlock(mutex_counter.mutex) == 0
+           && "Should be able to unlock the mutex after test execution");
+
+    return NULL;
+}
+
+void
+non_main_test(pthread_mutex_t *mutex, int num_iter)
+{
+    MutexCounter mutex_counter;
+    int counter = 0;
+    mutex_counter.mutex = mutex;
+    mutex_counter.counter = &counter;
+    mutex_counter.limit = num_iter;
+    mutex_counter.is_sleeping = false;
+
+    pthread_t tid = 0;
+    spawn_thread(&tid, inc_until_limit, &mutex_counter);
+
+    assert(tid != 0 && "TID can't be 0 after successful thread creation");
+    pthread_join(tid, NULL);
+    assert(*mutex_counter.counter == mutex_counter.limit);
+}
+
+void
+main_non_main_test(pthread_mutex_t *mutex, int num_iter)
+{
+    MutexCounter mutex_counter;
+    int counter = 0;
+    mutex_counter.mutex = mutex;
+    mutex_counter.counter = &counter;
+    mutex_counter.limit = num_iter;
+    mutex_counter.is_sleeping = false;
+
+    pthread_t tid = 0;
+    spawn_thread(&tid, inc_until_limit, &mutex_counter);
+
+    assert(tid != 0 && "TID can't be 0 after successful thread creation");
+    inc_until_limit(&mutex_counter);
```

Comment:
since `is_sleeping` is disabled, how likely is that the counter will reach the limit here, before another thread is spawned? Is num_iter large enough? Otherwise, we're actually not testing the concurrent updates.

Reply:
well here's the example for fast-interp

I think it's fine 

```
Starting concurrent_inc_non_sleep test
Thread calls count by TID
TID: 71316; Calls: 10044
TID: 104228; Calls: 8523
TID: 137140; Calls: 8725
TID: 170052; Calls: 6955
TID: 202964; Calls: 7502
TID: 235876; Calls: 6953
TID: 268788; Calls: 7106
TID: 301700; Calls: 7237
TID: 334612; Calls: 7666
TID: 367524; Calls: 7547
TID: 400436; Calls: 7195
TID: 433348; Calls: 7170
TID: 1028; Calls: 7377
```

Reply:
it's happened once in CI for AOT so I increased to 250k and it didn't slow the test much as it's quite fast itself